### ISSUE TITLE
ID3D11DeviceContext::IAGetVertexBuffers/IASetVertexBuffers does not take array parameters.

### ISSUE THREAD

**Issue by chyyran** [microsoft/windows-rs#1567]
Using [feb7d25](https://github.com/microsoft/windows-rs/commit/feb7d258fa8957945eabe1ce18e57795f211f2ee) after #1562 got merged, it seems like 

`ID3D11DeviceContext::IAGetVertexBuffers` still has signature

```rust
pub unsafe fn IAGetVertexBuffers(&self, startslot: u32, numbuffers: u32, ppvertexbuffers: *mut Option<ID3D11Buffer>, pstrides: *mut u32, poffsets: *mut u32);

// Same with IASetVertexBuffer

pub unsafe fn IASetVertexBuffers(&self, startslot: u32, numbuffers: u32, ppvertexbuffers: *const Option<ID3D11Buffer>, pstrides: *const u32, poffsets: *const u32);
```

Therefore code like this does not work

```rust
#[derive(Default)]
struct IABackup<'ctx> {
    vertex_buffers: [Option<ID3D11Buffer>; D3D11_IA_VERTEX_INPUT_RESOURCE_SLOT_COUNT as usize],
    vertex_buffer_strides: [u32; D3D11_IA_VERTEX_INPUT_RESOURCE_SLOT_COUNT as usize],
    vertex_buffer_offs: [u32; D3D11_IA_VERTEX_INPUT_RESOURCE_SLOT_COUNT as usize],
    _pd: PhantomData<&'ctx ()>
}

impl <'ctx> IABackup<'ctx> {
    unsafe fn backup(context: &'ctx ID3D11DeviceContext) {
        let mut backup : IABackup<'ctx> = Default::default();
        context.IAGetVertexBuffers(0, D3D11_IA_VERTEX_INPUT_RESOURCE_SLOT_COUNT, &mut backup.vertex_buffers,
                                   &mut backup.vertex_buffer_strides, &mut backup.vertex_buffer_offs);
    }
}
```

Following #1562, I expected signatures like

```rust
pub unsafe fn IAGetVertexBuffers(&self, startslot: u32, ppvertexbuffers: &mut [Option<ID3D11Buffer>], pstrides: &mut [u32], poffsets: &mut [u32]);

pub unsafe fn IASetVertexBuffers(&self, startslot: u32, ppvertexbuffers: &[Option<ID3D11Buffer>], pstrides: &[u32], poffsets: &[u32]);
```

It seems like this is the only associated function affected in `ID3D11DeviceContext` as everything else seems to take slices correctly.

The only issue I see with the above signatures is that the input/output slices must all be of the same size. I'm not sure if there are other cases like this still being projected with pointers, but the array size restrictions can/should be projected idiomatically with min const generics (MSRV 1.51)

```rust
pub unsafe fn IAGetVertexBuffers<const N: usize>(&self, startslot: u32, ppvertexbuffers: &mut [Option<ID3D11Buffer>; N], pstrides: &mut [u32; N], poffsets: &mut [u32; N]);

pub unsafe fn IASetVertexBuffers<const N: usize>(&self, startslot: u32, ppvertexbuffers: &[Option<ID3D11Buffer>; N], pstrides: &[u32; N], poffsets: &[u32; N]);
```

The documentation for [`IASetVertexBuffers`](https://docs.microsoft.com/en-us/windows/win32/api/d3d11/nf-d3d11-id3d11devicecontext-iasetvertexbuffers) doesn't specify what happens if the slice lengths are mismatched so it's highly likely unsound in that scenario. However since the functions are marked unsafe anyways, I don't think it's too big of a deal if adding support for const generics in `bindgen` is too much of an undertaking, or if you don't want to bump the MSRV that high.

**Comment by kennykerr** [comment]
This was discussed in the comments of #1562. If an array length parameter is shared and any one of the pointer parameters is optional, then the slice transformation is abandoned as ambiguous. 

https://github.com/microsoft/windows-rs/blob/feb7d258fa8957945eabe1ce18e57795f211f2ee/crates/libs/bindgen/src/methods.rs#L263-L310

**Comment by kennykerr** [comment]
But that reminds me that if the shared array parameters are required then we should ensure that they are all the same length and for that a const generic might be appropriate.

**Comment by chyyran** [comment]
If the issue is that the optional arrays are allowed to be NULL or otherwise size N where N is the shared length parameter would a projection like the following make sense?

```rust
pub unsafe fn IAGetVertexBuffers<const N: usize>(&self, startslot: u32, ppvertexbuffers: Option<&mut [Option<ID3D11Buffer>; N]>, pstrides: Option<&mut [u32; N]>, poffsets: Option<&mut [u32; N]>);

pub unsafe fn IASetVertexBuffers<const N: usize>(&self, startslot: u32, ppvertexbuffers: Option<&[Option<ID3D11Buffer>; N]>, pstrides: Option<&[u32; N]>, poffsets: Option<&[u32; N]>);

```

**Comment by kennykerr** [comment]
Thanks! I'll give that a try - that seems like a good approach at first glance.

**Comment by kennykerr** [comment]
I fixed https://github.com/microsoft/windows-rs/issues/1567#issuecomment-1050867598 in #1570 but was not able to resolve https://github.com/microsoft/windows-rs/issues/1567#issue-1150168104. Closing for now. Feel free to reopen or create a new issue if you think of something else.
